contracts/utils.go

```diff
@@ -4,85 +4,150 @@ import (
 	"bytes"
 	"errors"
 	"math/big"
+	"time"
 
-	abipkg "github.com/celo-org/celo-blockchain/accounts/abi"
+	"github.com/celo-org/celo-blockchain/accounts/abi"
 	"github.com/celo-org/celo-blockchain/common"
 	"github.com/celo-org/celo-blockchain/common/hexutil"
 	"github.com/celo-org/celo-blockchain/core/vm"
 	"github.com/celo-org/celo-blockchain/log"
+	"github.com/celo-org/celo-blockchain/metrics"
 )
 
-// systemCaller is the caller when the EVM is invoked from the within the blockchain system.
-var systemCaller = vm.AccountRef(common.HexToAddress("0x0"))
+// SystemCaller is the caller when the EVM is invoked from the within the blockchain system.
+var SystemCaller = common.HexToAddress("0x0")
 
-func StaticCallFromSystem(evm *vm.EVM, contractAddress common.Address, abi abipkg.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64) (uint64, error) {
-	staticCall := func(transactionData []byte) ([]byte, uint64, error) {
-		return evm.StaticCall(systemCaller, contractAddress, transactionData, gas)
-	}
+type Contract struct {
+	abi     *abi.ABI
+	address common.Address
+	caller  common.Address
+}
 
-	return handleABICall(evm, abi, funcName, args, returnObj, staticCall)
+type Backend interface {
+	Call(caller vm.ContractRef, addr common.Address, input []byte, gas uint64, value *big.Int) (ret []byte, leftOverGas uint64, err error)
+	StaticCall(caller vm.ContractRef, addr common.Address, input []byte, gas uint64) (ret []byte, leftOverGas uint64, err error)
+	StopGasMetering()
+	StartGasMetering()
+	GetStateDB() vm.StateDB
 }
 
-func CallFromSystem(evm *vm.EVM, contractAddress common.Address, abi abipkg.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int) (uint64, error) {
-	call := func(transactionData []byte) ([]byte, uint64, error) {
-		return evm.Call(systemCaller, contractAddress, transactionData, gas, value)
+func NewContract(contractAbi *abi.ABI, contractAddress, defaultCaller common.Address) *Contract {
+	return &Contract{
+		abi:     contractAbi,
+		address: contractAddress,
+		caller:  defaultCaller,
 	}
-	return handleABICall(evm, abi, funcName, args, returnObj, call)
 }
 
-var (
-	errorSig     = []byte{0x08, 0xc3, 0x79, 0xa0} // Keccak256("Error(string)")[:4]
-	abiString, _ = abipkg.NewType("string", "", nil)
-)
+type ExecOpts struct {
+	Backend Backend
+	MaxGas  uint64
+	Value   *big.Int
+}
+type QueryOpts struct {
+	Backend Backend
+	MaxGas  uint64
+}
 
-func unpackError(result []byte) (string, error) {
-	if len(result) < 4 || !bytes.Equal(result[:4], errorSig) {
-		return "<tx result not Error(string)>", errors.New("TX result not of type Error(string)")
+func (c *Contract) Execute(opts ExecOpts, result interface{}, method string, args ...interface{}) (uint64, error) {
+	defer c.meterExecutionTime(method)()
```

Comment:
Given how small the difference is between `Execute` and `Query` compared to how much is the same, I would favor undoing this duplication.  In some cases, I would agree duplication is better to make the control flow simpler, but in this case I don't think so.

There could be a single private method they both use which does all the stuff before and after the call itself, and receives an anonymous function for the call (which will in one case use Call and in the other use StaticCall.